packages/yew-router/src/router.rs

```diff
@@ -1,232 +1,121 @@
-    /// New render function
-    fn new<F: RenderFn<Router<SW, STATE>, SW> + 'static>(f: F) -> Self {
-        Render(Rc::new(f))
-    }
-}
-impl<STATE: RouterState, SW: Switch + Clone> Debug for Render<SW, STATE> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Render").finish()
+impl<T> Clone for RenderFn<T> {
+    fn clone(&self) -> Self {
+        Self(Rc::clone(&self.0))
     }
 }
 
-/// Redirection function that takes a route that didn't match any of the Switch variants,
-/// and converts it to a switch variant.
-pub trait RedirectFn<SW, STATE>: Fn(Route<STATE>) -> SW {}
-impl<T, SW, STATE> RedirectFn<SW, STATE> for T where T: Fn(Route<STATE>) -> SW {}
-/// Clonable Redirect function
-#[derive(Clone)]
-pub struct Redirect<SW: Switch + 'static, STATE: RouterState>(
-    pub(crate) Rc<dyn RedirectFn<SW, STATE>>,
-);
-impl<STATE: RouterState, SW: Switch + 'static> Redirect<SW, STATE> {
-    fn new<F: RedirectFn<SW, STATE> + 'static>(f: F) -> Self {
-        Redirect(Rc::new(f))
+impl<T> PartialEq for RenderFn<T> {
+    fn eq(&self, other: &Self) -> bool {
+        // https://github.com/rust-lang/rust-clippy/issues/6524
+        #[allow(clippy::vtable_address_comparisons)]
+        Rc::ptr_eq(&self.0, &other.0)
     }
 }
-impl<STATE: RouterState, SW: Switch> Debug for Redirect<SW, STATE> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Redirect").finish()
-    }
+
+/// Props for [`Router`]
+#[derive(Properties, Clone, PartialEq)]
+pub struct RouterProps<R: Clone> {
+    /// Callback which returns [`Html`] to be rendered for the current route.
+    pub render: RenderFn<R>,
 }
 
-/// Properties for Router.
-#[derive(Properties, Clone)]
-pub struct Props<STATE: RouterState, SW: Switch + Clone + 'static> {
-    /// Render function that takes a Switch and produces Html
-    pub render: Render<SW, STATE>,
-    /// Optional redirect function that will convert the route to a known switch variant if explicit matching fails.
-    /// This should mostly be used to handle 404s and redirection.
-    /// It is not strictly necessary as your Switch is capable of handling unknown routes using `#[at = "/{*:any}"]`.
-    #[prop_or_default]
-    pub redirect: Option<Redirect<SW, STATE>>,
+#[doc(hidden)]
+pub enum Msg {
+    ReRender,
 }
 
-impl<STATE: RouterState, SW: Switch + Clone> Debug for Props<STATE, SW> {
-    fn fmt(&self, f: &mut Formatter) -> Result<(), FmtError> {
-        f.debug_struct("Props").finish()
-    }
+/// The router component.
+///
+/// When a route can't be matched, it looks for the route with `not_found` attribute.
+/// If such a route is provided, it redirects to the specified route.
+/// Otherwise `html! {}` is rendered and a message is logged to console
+/// stating that no route can be matched.
+/// See the [crate level document][crate] for more information.
+pub struct Router<R: Routable + Clone + PartialEq + 'static> {
```

Comment:
Just one more thing. Could you use the new `add_onpush_listener` to register the listener for the router?
You just have to add a `route: Option<R>` member to the struct (which is initialized to `current_route()` in `create` and updated by the listener in `update`) and then start with `match self.route`

Reply:
Tried that, it doesn't work. On initial render, route is `None` so that adds complexity and breaks tests

Reply:
It really shouldn't be `None` the first time. Did you actually set the route to `current_route()` in `create`?

Reply:
No, I set route to `None` in create and then in `rendered`, I attached the listener which sent a message to update the route

Reply:
Yeah, that wouldn't work.
You no longer need `rendered`, just `create`, `update`, and `view`.
In `create` you attach the listener and get the current route (since the listener will only notify you for future updates) and in `update` you receive said future route updates.

Reply:
Okay, that fixed it